### ISSUE TITLE
pdfpc Segmentation fault; after recent brew updates

### ISSUE THREAD

**Issue by LecrisUT** [Homebrew/homebrew-core#51133]
# Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

Run the program

## What happened (include command output)

[gist-logs pdfpc](https://gist.github.com/4479a7712c32ae9e9e3aa9248d95a636)

Somewhere along the lines of brew updates broke the functionality of this program.
I've just tried to open a pdf with pdfpc, when I get the segmentation fault. The last few days the program worked just fine, and I've tried reinstalling all of the programs and the problem persists.

<details>
  <summary>Command output</summary>
  <pre>

  Segmentation fault: 11

  </pre>
</details>

## What you expected to happen

To not get a segmentation fault I guess.

## Step-by-step reproduction instructions (by running `brew install` commands)

Install formula with newest repo and run any pdf file

```
brew install pdfpc
pdfpc pdfpc-demo.pdf
```

[PDF file I tried](https://github.com/pdfpc/pdfpc/releases/download/v4.4.0/pdfpc-demo.pdf)

**Comment by Bo98** [comment]
This appears to be from the Poppler update from 0.85.0 to 0.86.1. Sounds like an upstream bug but not entirely sure where.

**Comment by LecrisUT** [comment]
@Bo98  I'm curious what black magic you used to find that out?

Confirmed that it indeed is the problem with `poppler`. I've reverted to 0.85 and works as a charm. Are there other formulas broken by it?

From the logs it's definitely weird that `poppler 0.86.1` uses the libraries of `poppler 0.85.0`

**Comment by Bo98** [comment]
> Are there other formulas broken by it?

None have been reported so far.

> From the logs it's definitely weird that `poppler 0.86.1` uses the libraries of `poppler 0.85.0`

Not sure what you mean by this.

**Comment by LecrisUT** [comment]
> > From the logs it's definitely weird that `poppler 0.86.1` uses the libraries of `poppler 0.85.0`
> 
> Not sure what you mean by this.

For example, I've found this in the gist-log
`POPPLER_VERSION:INTERNAL=0.85.0`

> But some of them, like `pdfpc`, have pretty poor tests that they are borderline useless.

I will try to make a better test for `pdfpc`. Are the automated tests done in brew able to emulate monitors?

**Comment by Bo98** [comment]
> For example, I've found this in the gist-log
> `POPPLER_VERSION:INTERNAL=0.85.0`

Your logs predate the Poppler update. If you reinstall from source you'll likely get an updated version in your logs. But ultimately, it won't fix the problem as I've tried recompiling `pdfpc` and I still get the same issue.

> Are the automated tests done in brew able to emulate monitors?

They are run on CI machines that have display capabilities (albeit they are VMs so there's some differences to a "real" machine). In this case, it would be sufficient enough for it to just open a PDF and close the program once it is done loading (which might be hard to detect so may involve a `sleep` call).

**Comment by LecrisUT** [comment]
Thanks for the tips I'll work on a test and make a PR to test it I guess. Is there a way to simulate the VM tests locally?

Do you think it's possible to track down the `poppler` problem? I'm clueless on how to debug that.

**Comment by Bo98** [comment]
> Is there a way to simulate the VM tests locally?

No. Just assume it is the same and we'll find out if it's not. There's no difference more often than not.

> Do you think it's possible to track down the `poppler` problem?

I don't really know at the moment to be honest. I'll maybe try some other formulae and see if it happens to anything else.

**Comment by Bo98** [comment]
> we'll find out if it's not

Well, we'll only properly find out when pdfpc is functional again.

**Comment by LecrisUT** [comment]
I'm confused about the `poppler` formula. Why does it have to have:
```
make install
make clean
```
before `ccmake`, `make`?
I've also tried to `--build-from-source`, but I get `Errno::ENOENT: No such file or directory - libpoppler.a`, even though when I make interactive with the exact same steps, all of the lib files are there.

I'm trying to get the backtrace for the poppler people to debug. I'm just now dipping my toes into `gdb` and all of that. Are there alternatives native to MacOS, or any differences between how linux and mac handle libraries that I should be aware of?
I only get '??' in gdb backtrace atm.

**Comment by LecrisUT** [comment]
The previous error only appears when I add `-DCMAKE_BUILD_TYPE=debug`, but the fact that `libpoppler.a` is present remains.
Building from source without any changes I get:
```
Error: An exception occurred within a child process:
  NameError: undefined local variable or method `ls' for #<Formulary::FormulaNamespacee6c6af297a63958129ae6bf75c7c4ab3::Poppler:0x00007f9afda166f8>
```

**Comment by Bo98** [comment]
> Why does it have to have:
> ```
> make install
> make clean
> ```
>
> before `ccmake`, `make`?

Looks like it's to install the shared libraries first and then the static libraries. Evidently they can't be installed in one build operation.

> I'm just now dipping my toes into `gdb` and all of that. Are there alternatives native to MacOS

`lldb` generally works better.

> ```
>   NameError: undefined local variable or method `ls'
> ```

`ls` does not appear in the current Poppler formula file in homebrew-core.

**Comment by LecrisUT** [comment]
> `lldb` generally works better.

Thank you for the hint it does work better. 
> > ```
> >   NameError: undefined local variable or method `ls'
> > ```
> 
> `ls` does not appear in the current Poppler formula file in homebrew-core.

My bad, my bad. I just saw a typo in my formula. It does compile fine. However why does building with `-DCMAKE_BUILD_TYPE=debug` flag fail to find `libpoppler.a`? I can clearly see the file if I manually make it with the exact same flags.

**Comment by Bo98** [comment]
> However why does building with `-DCMAKE_BUILD_TYPE=debug` flag fail to find `libpoppler.a`?

I don't know - I can't reproduce that issue on my end.

**Comment by LecrisUT** [comment]
The `-DCMAKE_BUILD_TYPE` was again my fault.
Note to self: Don't ignore the commas in the formulas.

It seems the `poppler 0.86.1` have started to appear elsewhere. In the end it doesn't seem that `lldb` backtrace is detailed enough even with `-DCMAKE_BUILD_TYPE=debugfull`. It seems that `poppler` has more tests available. I'll see if those offer any clue.

**Comment by Bo98** [comment]
macOS can store the debug symbols separate from the library (e.g. a dSYM folder).

You could install `--keep-tmp` and have a look around the build directory left behind in `/tmp/poppler-*`

**Comment by LecrisUT** [comment]
Hmm, does `lldb` automatically figure out those debug symbols, if not how do you link to them?

**Comment by Bo98** [comment]
`lldb` can automatically find them but if they were never installed then there's only so much `lldb` can do.

**Comment by LecrisUT** [comment]
I've confirmed that all of the `poppler`'s `make test` are successful.
Even with `--keep-tmp` I wasn't able to get more detailed debug info. Here is the backtrace (chopped down a bit for readability):
```
Process 55898 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x20)
    frame #0: 0x00000001006b69bc libpoppler-glib.8.dylib`_poppler_action_new(_PopplerDocument*, LinkAction const*, char const*) + 1369
libpoppler-glib.8.dylib`_poppler_action_new:
->  0x1006b69bc <+1369>: testb  $0x1, 0x20(%rax)
    0x1006b69c0 <+1373>: jne    0x1006b69c8               ; <+1381>
    0x1006b69c2 <+1375>: addq   $0x21, %rax
    0x1006b69c6 <+1379>: jmp    0x1006b69d5               ; <+1394>
Target 0: (pdfpc) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x20)
  * frame #0: 0x00000001006b69bc libpoppler-glib.8.dylib`_poppler_action_new(_PopplerDocument*, LinkAction const*, char const*) + 1369
    frame #1: 0x00000001006bd763 libpoppler-glib.8.dylib`poppler_page_get_link_mapping + 200
    frame #2: 0x000000010001ab92 pdfpc`pdfpc_metadata_pdf_get_action_mapping(self=<unavailable>, page_num=<unavailable>) at pdf.c:3163
    frame #3: 0x0000000100035021 pdfpc`_pdfpc_view_behaviour_pdf_link_on_entering_slide_pdfpc_view_pdf_entering_slide [inlined] block7_data_unref(_userdata_=<unavailable>) at pdf_link.c:569
```

The developers at `poppler` would want the line number where `_poppler_action_new` has failed at, not only the function name.

**Comment by Bo98** [comment]
@LecrisUT I have #51745 with a possible fix, based upon an upstream patch. Can you try it out and see if it works for you?

**Comment by LecrisUT** [comment]
Yeah, everything works perfectly. Thanks for all the support
